src/sctools/platform.py

```diff
@@ -629,3 +629,304 @@ def attach_barcodes(cls, args=None):
         cls._tag_bamfile(args.u2, args.output_bamfile, tag_generators)
 
         return 0
+
+
+class BarcodePlatform(GenericPlatform):
+    """Command Line Interface for 10x Genomics v2 RNA-sequencing programs
+
+    This class defines several methods that are created as CLI tools when sctools is installed
+    (see setup.py)
+
+    Attributes
+    ----------
+    cell_barcode : fastq.EmbeddedBarcode
+        A data class that defines the start and end position of the cell barcode and the tags to
+        assign the sequence and quality of the cell barcode
+    molecule_barcode : fastq.EmbeddedBarcode
+        A data class that defines the start and end position of the molecule barcode and the tags
+        to assign the sequence and quality of the molecule barcode
+    sample_barcode : fastq.EmbeddedBarcode
+        A data class that defines the start and end position of the sample barcode and the tags
+        to assign the sequence and quality of the sample barcode
+
+    Methods
+    -------
+    attach_barcodes()
+        Attach barcodes from the forward (r1) and optionally index (i1) fastq files to the reverse
+        (r2) bam file
+
+    """
+    cell_barcode = None
+    molecule_barcode = None
+    sample_barcode = None
+
+    @classmethod
+    def _validate_barcode_input(cls, given_value, min_value):
+        """Validates that the barcode input is greater than a min value
+
+        Parameters
+        ----------
+        given_value : int
+            the given value that must be greater than the min_value,
+            (barcode length or barcode starting position)
+
+        min_value : int
+            the min value that the given_value must be greater than
+
+        Returns
+        -------
+        given_value : int
+            return given value if valid
+
+        """
+        if given_value < min_value:
+            raise argparse.ArgumentTypeError('Invalid barcode length/position')
+        return given_value
+
+    @classmethod
+    def _validate_barcode_args(cls, args):
+        """Validates that the barcode start position is greater than 0
+
+        Parameters
+        ----------
+        args : Iterable[str]
+            arguments list, The default value of None, when passed to `parser.parse_args`
+            causes the parser to read `sys.argv`
+
+        Returns
+        -------
+        args : Iterable[str]
+            return arguments list if valid
+
+        """
+        # check that both the barcode length and position are given as arguments
+        # or that neither barcode length and position are given as arguments (XOR boolean logic)
+        cell_barcode_start_pos_exists = bool(args.cell_barcode_start_pos) or (args.cell_barcode_start_pos == 0)
```

Comment:
This functionality is repeated enough that I would write out a short function to replace it: 

```python
def _check_barcode_length_and_position(length: Union[int, str], position: int, name: str):
    """return True if both length and positions are provided, or neither are"""
    if isinstance(length, str):
        length = ast.literal_eval(length)
    if length =! bool(position): 
        raise argparse.ArgumentError(
            f"If either {name} barcode position or length are provided, both must be"
        )

```